components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/EmailDomainValidationHandler.java

```diff
@@ -0,0 +1,233 @@
+            if (authenticatorConfig == null) {
+                continue;
+            }
+
+            ApplicationAuthenticator authenticator = authenticatorConfig.getApplicationAuthenticator();
+            if (authenticator instanceof FederatedApplicationAuthenticator) {
+                Map<String, String> localClaimValues;
+                if (stepConfig.isSubjectAttributeStep()) {
+                    localClaimValues =
+                            (Map<String, String>) context.getProperty(FrameworkConstants.UNFILTERED_LOCAL_CLAIM_VALUES);
+                } else {
+                    /*
+                     * Need to validate even if this is not the subject attribute step since
+                     * jit provisioning will happen in both scenarios.
+                     */
+                    localClaimValues =
+                            FrameworkUtils.getLocalClaimValuesOfIDPInNonAttributeSelectionStep(context, stepConfig,
+                                    context.getExternalIdP());
+                }
+
+                Optional<String> emailDomain =
+                        extractEmailDomain(localClaimValues.get(FrameworkConstants.EMAIL_ADDRESS_CLAIM));
+
+                if (!emailDomain.isPresent()) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Email address not found or is not in the correct format." +
+                                " Email domain validation failed for tenant: " + context.getTenantDomain());
+                    }
+                    throw new PostAuthenticationFailedException(NO_EMAIL_ATTRIBUTE_FOUND.getCode(),
+                            NO_EMAIL_ATTRIBUTE_FOUND.getMessage());
+                }
+
+                if (!isValidEmailDomain(context, emailDomain.get())) {
+                    throw new PostAuthenticationFailedException(INVALID_EMAIL_DOMAIN.getCode(),
+                            String.format(INVALID_EMAIL_DOMAIN.getMessage(), context.getTenantDomain()));
+                }
+            }
+        }
+        return PostAuthnHandlerFlowStatus.SUCCESS_COMPLETED;
+    }
+
+    private boolean isEmailDomainDiscoveryEnabled(String primaryOrganizationId)
+            throws OrganizationConfigException, OrganizationManagementException {
+
+        String tenantDomain = FrameworkServiceDataHolder.getInstance().getOrganizationManager()
+                .resolveTenantDomain(primaryOrganizationId);
+
+        DiscoveryConfig discoveryConfiguration =
+                FrameworkServiceDataHolder.getInstance().getOrganizationConfigManager()
+                        .getDiscoveryConfigurationByTenantId(IdentityTenantUtil.getTenantId(tenantDomain));
+        List<ConfigProperty> configProperties = discoveryConfiguration.getConfigProperties();
+        for (ConfigProperty configProperty : configProperties) {
+            if (configProperty.getKey().equals(EMAIL_DOMAIN_ENABLE)) {
```

Comment:
```suggestion
            if (EMAIL_DOMAIN_ENABLE.equals(configProperty.getKey())) {
```
To avoid possible NPE